communities/src/lib.rs

```diff
@@ -264,50 +357,110 @@ impl<T: Config> Module<T> {
         Ok(())
     }
 
-    fn validate_locations(loc: &Vec<Location>, cids: &Vec<CommunityIdentifier>) -> DispatchResult {
-        ensure!(loc.len() <= 1000, <Error<T>>::TooManyLocations);
-        ensure!(!loc.is_empty(), <Error<T>>::TooFewLocations);
+    fn get_relevant_neighbor_buckets(geo_hash: &geohash::String, location: &Location)-> Result<Vec<geohash::String>, Error<T>> {
+        let mut relevant_neighbor_buckets: Vec<geohash::String> = Vec::new();
+        let neighbors = neighbors(&geo_hash).map_err(|_| <Error<T>>::InvalidGeohash)?;
+        let (bucket_center_lon, bucket_center_lat, bucket_lon_error, bucket_lat_error) = decode(&geo_hash).map_err(|_| <Error<T>>::InvalidGeohash)?;
+        let bucket_min_lat = bucket_center_lat - bucket_lat_error;
+        let bucket_max_lat = bucket_center_lat + bucket_lat_error;
+        let bucket_min_lon = bucket_center_lon - bucket_lon_error;
+        let bucket_max_lon = bucket_center_lon + bucket_lon_error;
 
-        for l1 in loc.iter() {
-            ensure!(Self::is_valid_geolocation(&l1), <Error<T>>::InvalidLocation);
-            //test within this communities' set
-            for l2 in loc.iter() {
-                if l2 == l1 {
-                    continue;
-                }
-                ensure!(
-                    Self::solar_trip_time(&l1, &l2) >= MIN_SOLAR_TRIP_TIME_S,
-                    <Error<T>>::MinimumDistanceViolationWithinCommunity
-                );
-            }
-            // prohibit proximity to poles
-            if Self::haversine_distance(&l1, &NORTH_POLE) < DATELINE_DISTANCE_M
-                || Self::haversine_distance(&l1, &SOUTH_POLE) < DATELINE_DISTANCE_M
-            {
-                warn!(target: LOG, "location too close to pole: {:?}", l1);
-                return Err(<Error<T>>::MinimumDistanceViolationToPole)?;
-            }
-            // prohibit proximity to dateline
-            let dateline_proxy = Location {
-                lat: l1.lat,
-                lon: DATELINE_LON,
-            };
-            if Self::haversine_distance(&l1, &dateline_proxy) < DATELINE_DISTANCE_M {
-                warn!(target: LOG, "location too close to dateline: {:?}", l1);
-                return Err(<Error<T>>::MinimumDistanceViolationToDateLine)?;
-            }
-            // test against all other communities globally
-            for other in cids.iter() {
-                for l2 in Self::locations(other) {
-                    if Self::solar_trip_time(&l1, &l2) < MIN_SOLAR_TRIP_TIME_S {
-                        warn!(target: LOG,
-                                     "location {:?} too close to previously registered location {:?} with cid {:?}",
-                                     l1, l2, other);
-                        return Err(<Error<T>>::MinimumDistanceViolationToOtherCommunity)?;
-                    }
-                }
+
+        //check if northern neighbour bucket needs to be included
```

Comment:
this block contains massively redundant code. But I guess the only way to reduce this would be a macro. Might not be worth bothering....